arduino/src/arduino.cpp

```diff
@@ -0,0 +1,225 @@
+{
+    if (checkConnection() != Arduino::ERROR_ARDUINO::SUCCSESS)
+    {
+        std::cout << "there was a problem with the connection" << std::endl;
+    }
+    else
+    {
+
+        std::cout << "the connection is successful!" << std::endl;
+    }
+}
+
+void Arduino::addFile(const std::string &file)
+{
+    this->files.push_back(file);
+}
+
+int Arduino::checkSum(const char *message)
+{
+    int sum = 0;
+    const char *p = &message[0];
+    while (*p != NULL)
+    {
+        sum += *message;
+        ++p;
+    }
+
+    return sum;
+}
+
+Arduino::ERROR_ARDUINO Arduino::openSerial()
+{
+    int err = serial.openDevice(this->portName.c_str(), this->buadRate);
+    if (err != 1)
+        return Arduino::ERROR_ARDUINO::ERROR_WITH_SERIALPORT;
+    this->isOpen = true;
+    return Arduino::ERROR_ARDUINO::SUCCSESS;
+}
+
+Arduino::ERROR_ARDUINO Arduino::serialCommend(const std::string &message)
+{
+    if (!this->isOpen)
+        return Arduino::ERROR_ARDUINO::ERROR_WITH_SERIALPORT;
+    serial.writeString(message.c_str());
+    // check sum should be added.
+    return Arduino::ERROR_ARDUINO::SUCCSESS;
+}
+
+Arduino::ERROR_ARDUINO Arduino::writeFromBufferToFile(const std::string &data, int place)
+{
+    std::ofstream dataLog;
+    std::string fileName = makeFileName(place);
+    dataLog.open(fileName, std::ios_base::app);
+    if (!dataLog.is_open())
+    {
+        return Arduino::ERROR_ARDUINO::ERROR_WITH_FILES;
+    }
+
+    dataLog << std::time(nullptr) << std::endl;
+    dataLog << data << std::endl;
+    dataLog.close();
+    return Arduino::ERROR_ARDUINO::SUCCSESS;
+}
+
+Arduino::ERROR_ARDUINO Arduino::getDataWithWhileLoop()
+{
+    Arduino::ERROR_ARDUINO return_code = Arduino::ERROR_ARDUINO::SUCCSESS;
+    return_code = openSerial();
+    if (return_code != Arduino::ERROR_ARDUINO::SUCCSESS)
+    {
+        goto l_cleanup;
+    }
+    while (true)
+    {
+        if (serial.available())
+        {
+            uint32_t command = receiveInt();
+            switch (command)
+            {
+            case Arduino::HEADERS::GET_FLOAT:
+                return_code = receiveFloatFromSensor();
+                if (return_code != Arduino::ERROR_ARDUINO::SUCCSESS)
+                    goto l_cleanup;
+                continue;
+            case Arduino::HEADERS::ASK_FOR_OVERALL_STATUS:
+                return_code = checkStatusFromArduino();
+                if (return_code != Arduino::ERROR_ARDUINO::SUCCSESS)
+                    goto l_cleanup;
+                continue;
+            default:
+                continue;
+            }
+        }
+    }
+
+l_cleanup:
+    return return_code;
+}
+
+Arduino::ERROR_ARDUINO Arduino::receiveFloatFromSensor()
+{
+    Arduino::ERROR_ARDUINO errorCode = Arduino::ERROR_ARDUINO::SUCCSESS;
+    int id = receiveInt();
+    float f = receiveFloat();
+    errorCode = this->writeFromBufferToFile(std::to_string(f), id);
+    return errorCode;
+}
+
+float Arduino::receiveFloat()
```

Comment:
if we are deciding the root of returning error code, then this should not return also float, simple because the value can be legitimately -1. Please change the function to return error code and reference the return value.

Reply:
fixed